docs/developer/designs/topology-awareness/README.md

```diff
@@ -0,0 +1,306 @@
+
+* Changes will be made to the allocation logic and/or a new plugin point will be created to get a possible group of nodes for a job that will be attempted and if not chosen a next one will be attempted.
+* This will work best if the first approach is implemented first to sort the nodes by their topology location and the available resources in each topology.
+* Significant performance reduction in big clusters is expected here.
+
+
+#### Topology Info struct
+
+Both approaches need to build a topology tree and update it with the changes the scheduler is doing during the session and simulation.
+We will read the Topology resources during the cluster session start phase and create TopologyInfo structs that will be maintained throughout the session with every change.
+
+```go
+// TopologyDomainID uniquely identifies a topology domain
+type TopologyDomainID string
+
+// TopologyInfo represents a topology tree for the cluster
+type TopologyInfo struct {
+    // Root of the topology tree
+    Root *TopologyDomainInfo
+
+    // Map of all domains by their ID for quick lookup
+    Domains map[TopologyDomainID]*TopologyDomainInfo
+
+    // Name of this topology configuration
+    Name string
+}
+
+// TopologyDomainInfo represents a node in the topology tree
+type TopologyDomainInfo struct {
+    // Unique ID of this domain
+    ID TopologyDomainID
+
+    // Name of this domain
+    Name string
+
+    // Level in the hierarchy (e.g., "datacenter", "zone", "rack", "node")
+    Level string
+
+    // Parent domain, nil for root
+    Parent *TopologyDomainInfo
+
+    // Child domains
+    Children []*TopologyDomainInfo
+
+    // Nodes that belong to this domain
+    Nodes map[string]*node_info.NodeInfo
+
+    // Total available resources in this domain
+    AvailableResources v1.ResourceList
+
+    // Total allocated resources in this domain
+    AllocatedResources v1.ResourceList
+
+    // Depth in the tree from root (0 for root)
+    Depth int
+}
+```
+
+There are two possibilities as to where the struct should be built and kept:
+  * Plugin - We will use a plugin that will build it in `OnSessionOpen` and will use hooks similar to DRA plugin to keep tabs on all the changes made in the cluster.
+  * ClusterInfo - This can be done with TopologyInfo structs in the ClusterInfo package, the plugin will later read this from the cache to make node order decision and will update this struct with simulated allocations.
+Because this might be resource intensive we will leave users with the option to totally disable it and only reference it in the plugin.
+
+#### Topology Awareness Plugin
+
+##### Setup
+Maintain topology tree with available resources using the Allocate/Deallocate event handlers (similar to the dynamicresources plugin)
+
+##### Pre Process
+Register the PrePredicate function (Possibly change its name to PreFilter ? ) to create a job-specific and look for allocation in that tree.
+
+####  Node Ordering and Filtering
```

Comment:
Do you think we can leverage FeasibleNodes and attempt multiple scenarios (each scenario with a different set of nodes) from the Pre-Process step?

Reply:
I think that we could use this to implement feasible nodes faster, but not the other way around.